lib/internal/process/esm_loader.js

```diff
@@ -34,14 +34,14 @@ exports.importModuleDynamicallyCallback = async function(wrap, specifier) {
   throw new ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING();
 };
 
-let ESMLoader = new Loader();
-exports.ESMLoader = ESMLoader;
```

Comment:
I'm not sure what the purpose of this first instance is—it appears to only exist to provide access to `esmLoader.import()`, used on line 56 below.

If the intention was to have the first instance be "public" and the second instance to be "private", it's not: the first instance is exposed (here), and then so is the second (on line 61).

If there's not a specific reason for doing this, I think this is better managed with a constructor (which is what I've refactored it to be now).

Reply:
Please add back the nested hooked loader pattern that was in place here.

The reasoning is that the first instance is the loader loader, and that the second instance is the loader hooked by that loader. This is necessary because loaders themselves are modules which might load things from node_modules. So we want to ensure the modules that are supposed to be running with the loader hooks always will use those hooks by creating a new nested loader. This is in case a dependency happened to be shared between the loader and the code running.

Reply:
This explanation would be good to add to the code as a comment.

Reply:
@guybedford Ahh, okay, thank you! This was one of the main questions I had for you. I saw in the blame you were the author, but wasn't sure why you did it (one of those "🤔 this is suspicious—there must be a reason…"). I think the previous implementation did not actually achieve that (because of what I described above)?

Is that actually necessary though? A node_module module imported by a custom loader won't get loaded through itself, but will that actually matter? I _think_ the only actual difference it will make is to create multiple caches of those node_modules (in scenarios where the custom loader and something else both import the same node_module module). I'm thinking the only cases where it will matter is when someone is trying to do something they shouldn't (I can't think of a specific naughty example, but my gut is saying whatever it is, probably "don't").

If not correct, I'm happy to update it, but could I get a little more explanation so I can include it in a code comment for future code explorers? (and to ensure I update it to achieve the correct behaviour)

@GeoffreyBooth incoming 😁

Reply:
> A node_module module imported by a custom loader won't get loaded through itself (the custom loader), but will that actually matter?

The problem is if a userland module then later on loads that same module from node_modules. In this case, the cached version that the custom loader used will be used, and that module won't go through the hooks. This is the feature / invariant we definitely need to keep here.

Reply:
Ahh, now I get it—an example where it would matter would be the module has local state. ex

```js
const counter = 0;

export default function doSomething() {
  // …
  ++counter;
}
```

Reply:
addressed both: 5327597